### PR TITLE
Issue #6207: Add Xpath Regression Test for ThrowsCount

### PR DESCRIPTION
**Pull request by AbhinavRajputEXE**
Issue #6207

### CODE REVIEW

**Comment by AbhinavRajputEXE** [comment]
@romani 
I know there are some errors in this but, I'm facing a few problems, as I'm doing XPath issue first time
I need some help

I made these changes created test files and input files for those tests and I'm getting this error
can you please help me😅

```
[ERROR] Tests run: 3, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 0.056 s <<< FAILURE! - in org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest
[ERROR] testOne  Time elapsed: 0.039 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
input file is expected to have a warning comment on line number 4
expected to be true
        at org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest.testOne(XpathRegressionThrowsCountTest.java:54)

[ERROR] testTwo  Time elapsed: 0.009 s  <<< FAILURE!
com.google.common.truth.ComparisonFailureWithFacts: 
error message 0
expected: ?30: Throws count is 5 (max allowed is 3).
but was : ?30: Throws count is 3 (max allowed is 2).
        at org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest.testTwo(XpathRegressionThrowsCountTest.java:75)

[ERROR] testThree  Time elapsed: 0.008 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
input file is expected to have a warning comment on line number 15
expected to be true
        at org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest.testThree(XpathRegressionThrowsCountTest.java:98)
```

in the "testOne" why it is expecting to "warn" message in line 
and also the getCheckMessage arguments are correct or not?

**Comment by nick-mancuso** [comment]
@AbhinavRajputEXE please take time to create a config file that uses ThrowsCount and execute Checkstyle via command line over the input files you’ve made to help you understand how the check works.

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package org.checkstyle.suppressionxpathfilter.throwscount;
+
+public class SuppressionXpathRegressionThrowsCount1 {
+    public void myFunction() throws CloneNotSupportedException,
+            ArrayIndexOutOfBoundsException,
+            StringIndexOutOfBoundsException,
+            IllegalStateException,
+            NullPointerException { // warn, max allowed is 4
```
Comment (baratali) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java:8]:
```
[ERROR] testOne  Time elapsed: 0.039 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
input file is expected to have a warning comment on line number 4
expected to be true
```
it expects a warning comment `// warn, max allowed is 4` on line 4, however, you put it on line 8. Such comments have to match with actual warnings.

**Review by baratali — CHANGES_REQUESTED** [review]

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java (hunk 1 above)
Comment (AbhinavRajputEXE) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java]: done

**Comment by AbhinavRajputEXE** [comment]
The error message was saying to place a warning message comment at line 15 instead of the line which is warned in the documentation example, so I placed a warning comment at line 15 but it still is giving me an error.
In the case of the third test for the third input file.

and the errors for the previous 2 tests and input files have been fixed.

```
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.055 s <<< FAILURE! - in org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest
[ERROR] testThree  Time elapsed: 0.042 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
input file is expected to have a warning comment on line number 15
expected to be true
        at org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest.testThree(XpathRegressionThrowsCountTest.java:104)

```

What am I doing wrong?

**Comment by baratali** [comment]
@AbhinavRajputEXE , it's because you specified wrong input file name in the third test, it should be SuppressionXpathRegressionThrowsCount**3**.java:
```java
    public void testThree() throws Exception {
        final File fileToProcess =
                new File(getPath("SuppressionXpathRegressionThrowsCount2.java"));
```

**Comment by AbhinavRajputEXE** [comment]
> @AbhinavRajputEXE , it's because you specified wrong input file name in the third test, it should be SuppressionXpathRegressionThrowsCount**3**.java:

oh sorry😅,
I was focusing on tweaking the input file and completely missed that.
thanks.

**Comment by romani** [comment]
@AbhinavRajputEXE, still red, please address failures

**Comment by AbhinavRajputEXE** [comment]
> @AbhinavRajputEXE, CI still red, please address failures

I tried to debug it was not able to find what is causing this error, 
this part.
```
expected: 0
but was : 1
```

```

[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.059 s <<< FAILURE! - in org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest
[ERROR] testThree  Time elapsed: 0.047 s  <<< FAILURE!
com.google.common.truth.ComparisonFailureWithFacts: 
unexpected output: /Users/abhinav/Documents/Checkstyle2/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount3.java:17:32: Throws count is 5 (max allowed is 4).
expected: 0
but was : 1
        at org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest.testThree(XpathRegressionThrowsCountTest.java:107)

```

**Comment by nick-mancuso** [comment]
> > @AbhinavRajputEXE, CI still red, please address failures
> 
> I tried to debug it was not able to find what is causing this error, this part.
> 
> ```
> expected: 0
> but was : 1
> ```
> 
>  ```
> 
> [ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.059 s <<< FAILURE! - in org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest
> [ERROR] testThree  Time elapsed: 0.047 s  <<< FAILURE!
> com.google.common.truth.ComparisonFailureWithFacts: 
> unexpected output: /Users/abhinav/Documents/Checkstyle2/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount3.java:17:32: Throws count is 5 (max allowed is 4).
> expected: 0
> but was : 1
>         at org.checkstyle.suppressionxpathfilter.XpathRegressionThrowsCountTest.testThree(XpathRegressionThrowsCountTest.java:107)
> ```

The number of XPath queries and violations should be the same.

**Comment by AbhinavRajputEXE** [comment]
> The number of XPath queries and violations should be the same.

I added the XpathQuery for the second privateFunc with a warning too
but now the error says it's only expecting it for the first "myFunction" function with warning
do I need to change something else too?
```

[ERROR] Failures: 
[ERROR]   XpathRegressionThrowsCountTest.testThree:112->AbstractXpathTestSupport.runVerifications:215->AbstractXpathTestSupport.verifyXpathQueries:107 Generated queries do not match expected ones
missing (1): /COMPILATION_UNIT/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionThrowsCount3']]/OBJBLOCK/METHOD_DEF[./IDENT[@text='privateFunc']]/LITERAL_THROWS[./IDENT[@text='CloneNotSupportedException']]
---
expected   : [/COMPILATION_UNIT/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionThrowsCount3']]/OBJBLOCK/METHOD_DEF[./IDENT[@text='myFunction']]/LITERAL_THROWS[./IDENT[@text='CloneNotSupportedException']], /COMPILATION_UNIT/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionThrowsCount3']]/OBJBLOCK/METHOD_DEF[./IDENT[@text='privateFunc']]/LITERAL_THROWS[./IDENT[@text='CloneNotSupportedException']]]
but was    : [/COMPILATION_UNIT/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionThrowsCount3']]/OBJBLOCK/METHOD_DEF[./IDENT[@text='myFunction']]/LITERAL_THROWS[./IDENT[@text='CloneNotSupportedException']]]
[INFO] 
[ERROR] Tests run: 469, Failures: 1, Errors: 0, Skipped: 0

```

**Comment by nick-mancuso** [comment]
@AbhinavRajputEXE I am not sure that we fully support multiple violations across different XPaths in this test:

```
➜  suppressionxpathfilter git:(master) ✗ diff <(grep -cR "getCheckMessage(") <(grep -cR "@Test")
23c23
< XpathRegressionUnnecessarySemicolonInTryWithResourcesTest.java:3
---
> XpathRegressionUnnecessarySemicolonInTryWithResourcesTest.java:2
107c107
< XpathRegressionAvoidNestedBlocksTest.java:7
---
> XpathRegressionAvoidNestedBlocksTest.java:4


```
Please change test to have single violation.

**Comment by romani** [comment]
Single only plea

**Comment by AbhinavRajputEXE** [comment]
The third test was all about ignoring private methods for warnings, so I only kept the private method in the third input file and removed the previous public method, and for the third test "testThree" in the test file.
Now it's generating an Xpath query for one violation message per input file and test.
Is it fine?

**Comment by romani** [comment]
@AbhinavRajputEXE , you did it.
please send one more PR (for other Check) to confirm that you can fix such issues without problems.

**Comment by AbhinavRajputEXE** [comment]
>@AbhinavRajputEXE , you did it.
please send one more PR (for other Check) to confirm that you can fix such issues without problems.

thanks, sir

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package org.checkstyle.suppressionxpathfilter.throwscount;
+
+public class SuppressionXpathRegressionThrowsCount2 {
+    public void myFunction() throws IllegalStateException, // warn, max allowed is 2
+            ArrayIndexOutOfBoundsException,
```
Comment (romani) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java:5]: this Input is identical to previous.
No much value in it, please do violation on methods of interface at least.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java (hunk 2 above)
Comment (AbhinavRajputEXE) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java:5]: Should I add this method under an interface?
like this: 
```
interface myInterface {
        public void myFunction() throws IllegalStateException, // warn, max allowed is 2
                ArrayIndexOutOfBoundsException,
                NullPointerException;
    }
```
as this method doesn't have a body too.
Xpath query is also different now from the previous method:
```
"/COMPILATION_UNIT"
          + "/CLASS_DEF[./IDENT[@text='SuppressionXpathRegressionThrowsCount2']]"
          + "/OBJBLOCK/INTERFACE_DEF[./IDENT[@text='myInterface']]"
          + "/OBJBLOCK/METHOD_DEF[./IDENT[@text='myFunction']]"
          + "/LITERAL_THROWS[./IDENT[@text='IllegalStateException']]"
```

Comment (romani) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java]: Yes please

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package org.checkstyle.suppressionxpathfilter.throwscount;
+
+public class SuppressionXpathRegressionThrowsCount2 {
+    interface myInterface {
```
Comment (romani) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount2.java:4]: Make `interface SuppressionXpathRegressionThrowsCount2` as top level

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount3.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package org.checkstyle.suppressionxpathfilter.throwscount;
+
+public class SuppressionXpathRegressionThrowsCount3 {
+
+    public void myFunc() throws ArithmeticException,
+            NumberFormatException { // ok
```
Comment (romani) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount3.java:6]: This is again the same as first.
Please make violation on method that is defined in anonymous class.

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount3.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    private void privateFunc() throws CloneNotSupportedException, // warn, max allowed is 4
+            ClassNotFoundException,
+            IllegalAccessException,
+            ArithmeticException,
+            ClassCastException {
+        // body
+    }
+
+    private void func() throws IllegalStateException,
+            NullPointerException { // ok
```
Comment (romani) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount3.java:19]: Remove all code without warn. We test only xpath, all other code is not required

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by romani** [comment]
Always squash commits in one. If you do extra changes ammend existing commit, do not create new commit

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package org.checkstyle.suppressionxpathfilter.throwscount;
+
+interface SuppressionXpathRegressionThrowsCount1 {
```
Comment (romani) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java:3]: First input file should be class, second input file is an interface.

Third input class should be class with method, inside method anonymous class with method, and such method in annonimous class should be with violation

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java (hunk 6 above)
Comment (AbhinavRajputEXE) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java:3]: means the second input is correct, right?
I have to change 1st and 3rd.

Comment (romani) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java:3]: Yes

Comment (AbhinavRajputEXE) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/throwscount/SuppressionXpathRegressionThrowsCount1.java]: done

**Review by romani — APPROVED** [review]

**Comment by romani** [comment]
@AbhinavRajputEXE
I hope you are ready for second PR for xpath

**Comment by AbhinavRajputEXE** [comment]
> @AbhinavRajputEXE
> I hope you are ready for second PR for xpath 

Yes sir, I'm absolutely ready 🙂
I'll make a PR for it in the morning here.

**Review by baratali — APPROVED** [review]

**Pull request merged**